posts/2021/08/github-actions-benchmarks.md

```diff
@@ -0,0 +1,369 @@
+with colored patches for easier visual analysis. In the Y axis, we plot the performance ratio. Each
+of those vertical clouds include 75 points, one per benchmark test. Ideally, they should all fall at
+`y=1`. Of course, not all of them are there, but a surprising amount of them do!
+
+But! We do see some bigger deviations at certain times! How is that possible? Well, that's the error we were talking about!
+These outliers are the ones that would concern us because they can be interpreted as **false positives**.
+In other words: `asv` would report a performance regression, when in fact there's none. However,
+in the observed measurements, the outliers were always within `y ∈ (0.5, 1.4)`.
+That means we can affirm that the method is sensitive enough to detect performance regressions of 50% or
+more! This is good enough for our project and, in fact, some projects might even be happy with a threshold of `2.0`.
+
+> If you are curious about how we automatically downloaded the artifacts, parsed the output, and plotted
+> the performance ratios, check the Jupyter Notebook [here](https://gist.github.com/jaimergp/aa4f059c14e394c4089b320cb8b51b1a).
+
+# Can we make it run faster without losing accuracy?
+
+We just found out that this approach is sensitive enough _but_ it takes two hours to run. Is it possible
+to reduce that time without sacrificing sensitivity? We need to remember that `asv` is running several
+passes and repeats to reduce the measurement error, but maybe some of those default counter-measures
+are not needed. Namely:
+
+- The benchmark runs with `--interleave-processes`, but it can be disabled with `--no-interleave-processes`.
+The help message for this flag says:
+
+    > Interleave benchmarks with multiple processes across commits. This can avoid measurement biases
+    > from commit ordering, can take longer.
+
+    How much longer? Does it help keep error under control? We should measure that.
+
+- By default, all tests are run several times, with different schedules. There are two
+[benchmark attributes](https://asv.readthedocs.io/en/stable/benchmarks.html#timing-benchmarks)
+that govern these settings: `processes` and `repeat`. `processes` defaults to `2`, which means that
+the full suite will be run twice per commit. If we only do one pass (`processes=1`), we will reduce
+the running time in half, but will we lose too much accuracy?
+
+To answer both questions, we added more entries to the data collection workflow shown above by
+[parameterizing the `asv` command-line options](https://github.com/jaimergp/scikit-image/blob/e561996/.github/workflows/benchmarks-cron.yml#L14-L23).
+
+These are the results!
+
+## No process interleaving
+
+Disabling process interleaving should be faster and maybe the accuracy loss is not that bad. But...
+how bad? Here are the results:
+
+![Reliability of benchmarks in GitHub Actions, no interleaving. This 2D plots shows a 16-day timeseries in the X axis.
```

Comment:
Because you are a) comparing similar plots and (correctly) repeating alt text and b) relying heavily on the image to tell the story instead of summarizing it elsewhere, I recommend moving the intended takeaway of the plot to earlier in the alt text like:

```
Reliability of benchmarks in GitHub Actions, no interleaving. One particular 
outlier happened on the first Saturday with half the cloud below 0.75 performance. 
This 2D plots shows a 16-day timeseries in the X axis.
Each data point in the X axis corresponds to a cloud of 75 measurements (one per benchmark test).
The y-axis spread of each cloud corresponds to the performance ratio. Ideal measurements would have
a performance ratio of 1.0, since both runs returned the exact same performance. In practice this
does not happen. With no interleaving, the vertical spread is more evident, with several clouds
spreading beyond the desired interval.
```

While this doesn't quite follow the recommendations I posted, I think it makes more sense for screen reader users (and maybe people with cognitive disabilities) to be able to get the info they need most at the start and feel like they can skip the parts that are repeated later if they want.